### ISSUE TITLE
"core" file driver slightly broken in 3.3.0 with `backing_store=False`

### ISSUE THREAD

**Issue by paulmueller** [h5py/h5py#1916]
The usage of the "core" driver is not consistent from version 3.2.1 to 3.3.0.

While this worked in 3.2.1:

```python
import h5py
h5 = h5py.File(name="nonexistentfile.h5", driver="core", backing_store=False, mode="a")
```

... it doesn't in 3.3.0:

```
 /opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/h5py/_hl/files.py:447: in __init__
    swmr=swmr)
/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/h5py/_hl/files.py:211: in make_fid
    fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   OSError: Unable to open file (unable to open file)

h5py/h5f.pyx:100: OSError
```

 In 3.3.0, you have to specify `mode="w"` (which also works in 3.2.1).

```python
import h5py
h5 = h5py.File(name="nonexistentfile.h5", driver="core", backing_store=False, mode="w")
```

Note that the file "nonexistentfile.h5" does not exist.

The [docs](https://docs.h5py.org/en/stable/high/file.html#file-drivers) are not clear as to why `mode="a"` should not work.

Additional information
----------------------------

 * Operating System: Ubuntu 20.04
 * Python version: 3.8.5
 * Where Python was acquired: apt
 * h5py version: 3.3.0
 * HDF5 version: 1.12.0

```
Summary of the h5py configuration
---------------------------------

h5py    3.2.1
HDF5    1.12.0
Python  3.8.5 (default, May 27 2021, 13:30:53) 
[GCC 9.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.20.2
cython (built with) 0.29.22
numpy (built against) 1.17.5
HDF5 (built against) 1.12.0
```

**Comment by aragilar** [comment]
@paulmueller Can you confirm the issue isn't with different HDF5 versions?

If this is not due to different HDF5 versions, then https://github.com/h5py/h5py/commit/13eab57885ebf4a352b3129cea5d514bc9623137 seems like the only change that could cause this. The test would be to see what happens on 3.2.1 and 3.3.0 with the mode being `'r+'` and `'x'` (`'r+'` should fail and `'x'` should succeed). Then the question is why isn't the error `FileNotFoundError`.

**Comment by paulmueller** [comment]
3.2.1 r+: fails
3.2.1 x: works

3.3.0 r+: fails
3.3.0 x: works


Except for line numbers, the error messages are identical:

```
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-2-0bc1afcac632> in <module>
----> 1 a = h5py.File(name="test.h5", driver="core", backing_store=False, mode="r+")

~/repos/CellReel/.env/lib/python3.8/site-packages/h5py/_hl/files.py in __init__(self, name, mode, driver, libver, userblock_size, swmr, rdcc_nslots, rdcc_nbytes, rdcc_w0, track_order, fs_strategy, fs_persist, fs_threshold, **kwds)
    442             with phil:
    443                 fapl = make_fapl(driver, libver, rdcc_nslots, rdcc_nbytes, rdcc_w0, **kwds)
--> 444                 fid = make_fid(name, mode, userblock_size,
    445                                fapl, fcpl=make_fcpl(track_order=track_order, fs_strategy=fs_strategy,
    446                                fs_persist=fs_persist, fs_threshold=fs_threshold),

~/repos/CellReel/.env/lib/python3.8/site-packages/h5py/_hl/files.py in make_fid(name, mode, userblock_size, fapl, fcpl, swmr)
    199         fid = h5f.open(name, flags, fapl=fapl)
    200     elif mode == 'r+':
--> 201         fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)
    202     elif mode in ['w-', 'x']:
    203         fid = h5f.create(name, h5f.ACC_EXCL, fapl=fapl, fcpl=fcpl)

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/h5f.pyx in h5py.h5f.open()

OSError: Unable to open file (unable to open file)
```

**Comment by paulmueller** [comment]
And here once more a more elaborate error message with `mode="a"` from my local ipython:

```
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-4-9bb1716fb441> in <module>
----> 1 a = h5py.File(name="test.h5", driver="core", backing_store=False, mode="a")

~/repos/CellReel/.env/lib/python3.8/site-packages/h5py/_hl/files.py in __init__(self, name, mode, driver, libver, userblock_size, swmr, rdcc_nslots, rdcc_nbytes, rdcc_w0, track_order, fs_strategy, fs_persist, fs_threshold, **kwds)
    442             with phil:
    443                 fapl = make_fapl(driver, libver, rdcc_nslots, rdcc_nbytes, rdcc_w0, **kwds)
--> 444                 fid = make_fid(name, mode, userblock_size,
    445                                fapl, fcpl=make_fcpl(track_order=track_order, fs_strategy=fs_strategy,
    446                                fs_persist=fs_persist, fs_threshold=fs_threshold),

~/repos/CellReel/.env/lib/python3.8/site-packages/h5py/_hl/files.py in make_fid(name, mode, userblock_size, fapl, fcpl, swmr)
    209         # existing one (ACC_EXCL)
    210         try:
--> 211             fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)
    212         except FileNotFoundError:
    213             fid = h5f.create(name, h5f.ACC_EXCL, fapl=fapl, fcpl=fcpl)

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/h5f.pyx in h5py.h5f.open()

OSError: Unable to open file (unable to open file)
```

**Comment by takluyver** [comment]
Hmm, sorry about that. I think #1902 is the culprit, as @aragilar said. We want to catch the more specific error, but that depends on  HDF5 giving us the POSIX errno (`errno = 2`), and in this case the 'core' file driver doesn't ([code is here](https://github.com/HDFGroup/hdf5/blob/257c8c745446ee614f92624b30cec41d8d771e05/src/H5FDcore.c#L788-L789) - `HSYS_GOTO_ERROR` includes the errno, but `HGOTO_ERROR` doesn't).

@epourmal any chance this could be changed in HDF5?

If you need a workaround to do it like `mode='a'`, you can do something like this:

```python
try:
    h5 = h5py.File(filename, driver="core", backing_store=False, mode="r+")
except OSError:
    h5 = h5py.File(filename, driver="core", backing_store=False, mode="x")
```

(As you can see from the tracebacks, this is essentially how `mode='a'` is implemented)

**Comment by takluyver** [comment]
Closed by #1922
